src/Analyzer/Passes/IfTransformStringsToEnumPass.cpp

```diff
@@ -0,0 +1,306 @@
+        }
+
+        FindUsedFunctionsVisitor visitor(used_functions, function_names, stack_size + 1);
+        visitor.visit(function_node->getArgumentsNode());
+    }
+
+private:
+    /// we store only function aliases which are not modified, including the nodes owning it
+    /// we will modify only argument nodes
+    std::unordered_set<std::string_view> & used_functions;
+    const std::unordered_set<std::string> & function_names;
+    size_t stack_size;
+};
+
+template <typename EnumType>
+std::string makeStringsEnum(const EnumType & enum_type)
+{
+    std::string enum_string;
+
+    if constexpr (std::same_as<typename EnumType::FieldType, Int16>)
+    {
+        enum_string = "Enum16(";
+    }
+    else
+    {
+        static_assert(std::same_as<typename EnumType::FieldType, Int8>, "Invalid field type for enum");
+        enum_string = "Enum8(";
+    }
+
+    const auto & values = enum_type.getValues();
+    for (const auto & [value_string, value_index] : values)
+    {
+        enum_string += "\'" + value_string + "\' = " + std::to_string(value_index);
+
+        assert(value_index > 0);
+        if (static_cast<size_t>(value_index) < values.size())
+            enum_string += ", ";
+    }
+
+    enum_string += ")";
+
+    return enum_string;
+}
+
+/// We place strings in ascending order here under the assumption it colud speed up String to Enum conversion.
+template <typename EnumType>
+auto getDataEnumType(const std::set<std::string> & string_values)
+{
+    using EnumValues = typename EnumType::Values;
+    EnumValues enum_values;
+    enum_values.reserve(string_values.size());
+
+    size_t number = 1;
+    for (const auto & value : string_values)
+        enum_values.emplace_back(value, number++);
+
+    return std::make_shared<EnumType>(std::move(enum_values));
+}
+
+std::pair<DataTypePtr, std::string> getEnumTypeAndString(const std::set<std::string> & string_values)
+{
+    DataTypePtr result_type = nullptr;
+    std::string enum_string;
+    if (string_values.size() >= 255)
+    {
+        auto enum_type = getDataEnumType<DataTypeEnum16>(string_values);
+        enum_string = makeStringsEnum(*enum_type);
+        result_type = std::move(enum_type);
+    }
+    else
+    {
+        auto enum_type = getDataEnumType<DataTypeEnum8>(string_values);
+        enum_string = makeStringsEnum(*enum_type);
+        result_type = std::move(enum_type);
+    }
+
+    return {std::move(result_type), std::move(enum_string)};
+}
+
+void changeIfArguments(
+    QueryTreeNodePtr & first, QueryTreeNodePtr & second, const std::set<std::string> & string_values, const ContextPtr & context)
+{
+    auto [result_type, enum_string] = getEnumTypeAndString(string_values);
+    auto enum_literal = std::make_shared<ConstantValue>(enum_string, std::make_shared<DataTypeString>());
+    auto enum_literal_node = std::make_shared<ConstantNode>(std::move(enum_literal));
+
+    const auto create_cast_function = [&, &result_type = result_type](QueryTreeNodePtr & node)
```

Comment:
Seems that this can be reused as separate function that `changeIfArguments` and `changeTransformArguments` can use.